File: apps/gateway/src/openfang/sse.test.ts
import { describe, expect, it } from "bun:test";
import { OpenFangError } from "~/openfang/errors";
import { type OpenFangStreamEvent, parseFrame, parseOpenFangStream } from "~/openfang/sse";
import { streamFromChunks } from "~/openfang/test-support";

async function collect(chunks: string[]): Promise<OpenFangStreamEvent[]> {
  const events: OpenFangStreamEvent[] = [];
  for await (const event of parseOpenFangStream(streamFromChunks(chunks))) {
    events.push(event);
  }
  return events;
}

describe("parseFrame", () => {
  it("joins multiple data lines and strips the leading space", () => {
    const frame = parseFrame("event: chunk\ndata: line1\ndata: line2");
    expect(frame).toEqual({ event: "chunk", data: "line1\nline2" });
  });

  it("ignores comment lines and returns undefined for a keep-alive frame", () => {
    expect(parseFrame(": keep-alive")).toBeUndefined();
    expect(parseFrame("data: only-data-no-event")).toBeUndefined();
  });
});

describe("parseOpenFangStream", () => {
  it("projects the five known events into the typed union", async () => {
    const events = await collect([
      'event: phase\ndata: {"phase":"start","detail":null}\n\n',
      'event: chunk\ndata: {"content":"Hello ","done":false}\n\n',
      'event: tool_use\ndata: {"tool":"file_read"}\n\n',
      'event: tool_result\ndata: {"tool":"file_read","input":{"path":"a.ts"}}\n\n',
      'event: chunk\ndata: {"content":"world"}\n\n',
      'event: done\ndata: {"done":true,"usage":{"input_tokens":12,"output_tokens":3}}\n\n',
    ]);

    expect(events).toEqual([
      { type: "phase", phase: "start", detail: null },
      { type: "chunk", content: "Hello " },
      { type: "tool_use", tool: "file_read" },
      { type: "tool_result", tool: "file_read", input: { path: "a.ts" } },
      { type: "chunk", content: "world" },
      { type: "done", usage: { input_tokens: 12, output_tokens: 3 } },
    ]);
  });

  it("reassembles events split across chunk boundaries", async () => {
    const events = await collect([
      "event: chu",
      'nk\ndata: {"con',
      'tent":"hi"}\n\nevent: done\ndata: {"usage":{"input_tokens":1,"output_tokens":1}}\n\n',
    ]);
    expect(events).toEqual([
      { type: "chunk", content: "hi" },
      { type: "done", usage: { input_tokens: 1, output_tokens: 1 } },
    ]);
  });

  it("drops comment/keep-alive frames between events", async () => {
    const events = await collect([
      ": keep-alive\n\n",
      'event: chunk\ndata: {"content":"x"}\n\n',
      ": another comment\n\n",
      'event: done\ndata: {"usage":{"input_tokens":0,"output_tokens":0}}\n\n',
    ]);
    expect(events.map((e) => e.type)).toEqual(["chunk", "done"]);
  });

  it("preserves unknown named events as raw", async () => {
    const events = await collect([
      'event: heartbeat\ndata: {"beat":1}\n\n',
      'event: done\ndata: {"usage":{"input_tokens":0,"output_tokens":0}}\n\n',
    ]);
    expect(events[0]).toEqual({ type: "raw", event: "heartbeat", data: { beat: 1 } });
  });

  it("keeps reading after done and returns when the body ends", async () => {
    // The pinned upstream emits one `done` per model turn on the same
    // connection; hanging up at the first one lost the agent's final report.
    const events = await collect([
      'event: chunk\ndata: {"content":"I will look into it"}\n\n',
      'event: done\ndata: {"usage":{"input_tokens":10,"output_tokens":1}}\n\n',
      'event: phase\ndata: {"phase":"tool_loop","detail":null}\n\n',
      'event: tool_use\ndata: {"tool":"file_read"}\n\n',
      'event: tool_result\ndata: {"tool":"file_read","input":{}}\n\n',
      'event: done\ndata: {"usage":{"input_tokens":20,"output_tokens":2}}\n\n',
      'event: chunk\ndata: {"content":"Final report"}\n\n',
      'event: done\ndata: {"usage":{"input_tokens":30,"output_tokens":3}}\n\n',
      'event: phase\ndata: {"phase":"done","detail":null}\n\n',
    ]);
    expect(events.map((e) => e.type)).toEqual([
      "chunk",
      "done",
      "phase",
      "tool_use",
      "tool_result",
      "done",
      "chunk",
      "done",
      "phase",
    ]);
    expect(events[6]).toEqual({ type: "chunk", content: "Final report" });
    expect(events[7]).toEqual({ type: "done", usage: { input_tokens: 30, output_tokens: 3 } });
  });

  it("delivers text that follows a done instead of dropping it", async () => {
    // A trailing turn that never got its own `done` is still the last thing
    // the agent said; the consumer decides what to make of it.
    const events = await collect([
      'event: done\ndata: {"usage":{"input_tokens":1,"output_tokens":1}}\n\n',
      'event: chunk\ndata: {"content":"late"}\n\n',
    ]);
    expect(events).toEqual([
      { type: "done", usage: { input_tokens: 1, output_tokens: 1 } },
      { type: "chunk", content: "late" },
    ]);
  });

  it("marks the stream interrupted when it ends before any done event", async () => {
    const stream = parseOpenFangStream(
      streamFromChunks(['event: chunk\ndata: {"content":"partial"}\n\n']),
    );
    const seen: OpenFangStreamEvent[] = [];
    let error: unknown;
    try {
      for await (const event of stream) seen.push(event);
    } catch (caught) {
      error = caught;
    }
    expect(seen).toEqual([{ type: "chunk", content: "partial" }]);
    expect(error).toBeInstanceOf(OpenFangError);
    expect((error as OpenFangError).code).toBe("STREAM_INTERRUPTED");
  });

  it("marks the stream interrupted on malformed JSON, after prior events", async () => {
    const stream = parseOpenFangStream(
      streamFromChunks([
        'event: chunk\ndata: {"content":"ok"}\n\n',
        "event: chunk\ndata: {not json}\n\n",
      ]),
    );
    const seen: OpenFangStreamEvent[] = [];
    let error: unknown;
    try {
      for await (const event of stream) seen.push(event);
    } catch (caught) {
      error = caught;
    }
    expect(seen).toEqual([{ type: "chunk", content: "ok" }]);
    expect((error as OpenFangError).code).toBe("STREAM_INTERRUPTED");
  });
});

File: server/internal/openfang/sse_test.go
package openfang

import (
	"context"
	"errors"
	"io"
	"strings"
	"testing"
)

func TestMessageStreamProjectsPinnedEventsAndDone(t *testing.T) {
	t.Parallel()
	body := strings.Join([]string{
		": keep-alive\n\n",
		"event: phase\ndata: {\"phase\":\"start\",\"detail\":null}\n\n",
		"event: chunk\ndata: {\"content\":\"hello\",\"done\":false}\n\n",
		"event: tool_use\ndata: {\"tool\":\"file_read\"}\n\n",
		"event: tool_result\ndata: {\"tool\":\"file_read\",\"input\":{\"path\":\"a.go\"}}\n\n",
		"event: done\ndata: {\"done\":true,\"usage\":{\"input_tokens\":12,\"output_tokens\":3}}\n\n",
	}, "")
	stream := testMessageStream(body, 1024*1024, 128*1024)
	defer stream.Close()

	var got []StreamEvent
	for {
		event, err := stream.Next()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			t.Fatalf("Next() error = %v", err)
		}
		got = append(got, event)
	}
	if len(got) != 5 {
		t.Fatalf("event count = %d, want 5", len(got))
	}
	if got[0].Type != EventPhase || got[1].Content != "hello" ||
		got[2].Tool != "file_read" || string(got[3].Input) != `{"path":"a.go"}` ||
		got[4].Usage.InputTokens != 12 || got[4].Usage.OutputTokens != 3 {
		t.Fatalf("events = %#v", got)
	}
}

func TestMessageStreamEOFBeforeDoneIsInterruptedAfterPartialEvents(t *testing.T) {
	t.Parallel()
	stream := testMessageStream(
		"event: chunk\ndata: {\"content\":\"partial\"}\n\n",
		1024*1024,
		128*1024,
	)
	defer stream.Close()
	event, err := stream.Next()
	if err != nil || event.Type != EventChunk {
		t.Fatalf("first Next() event=%#v err=%v", event, err)
	}
	_, err = stream.Next()
	if !errors.Is(err, ErrStreamInterrupted) {
		t.Fatalf("second Next() error = %v, want ErrStreamInterrupted", err)
	}
}

// The pinned upstream emits done at the end of every model turn on the same
// connection. Reading must continue past it, and only the end of the body
// after at least one done is a clean finish.
func TestMessageStreamTreatsDoneAsTurnBoundaryUntilEOF(t *testing.T) {
	t.Parallel()
	body := strings.Join([]string{
		"event: phase\ndata: {\"phase\":\"start\",\"detail\":null}\n\n",
		"event: chunk\ndata: {\"content\":\"I'll look into it\"}\n\n",
		"event: tool_use\ndata: {\"tool\":\"file_read\"}\n\n",
		"event: tool_result\ndata: {\"tool\":\"file_read\",\"input\":{}}\n\n",
		"event: done\ndata: {\"done\":true,\"usage\":{\"input_tokens\":10,\"output_tokens\":1}}\n\n",
		"event: phase\ndata: {\"phase\":\"tool_loop\",\"detail\":null}\n\n",
		"event: chunk\ndata: {\"content\":\"Final report\"}\n\n",
		"event: done\ndata: {\"done\":true,\"usage\":{\"input_tokens\":20,\"output_tokens\":2}}\n\n",
		"event: phase\ndata: {\"phase\":\"done\",\"detail\":null}\n\n",
	}, "")
	stream := testMessageStream(body, 1024*1024, 128*1024)
	defer stream.Close()

	var got []StreamEvent
	for {
		event, err := stream.Next()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			t.Fatalf("Next() error = %v after %d events", err, len(got))
		}
		got = append(got, event)
	}
	if len(got) != 9 {
		t.Fatalf("event count = %d, want 9", len(got))
	}
	if got[4].Type != EventDone || got[4].Usage.InputTokens != 10 ||
		got[6].Content != "Final report" ||
		got[7].Type != EventDone || got[7].Usage.OutputTokens != 2 ||
		got[8].Type != EventPhase || got[8].Phase != "done" {
		t.Fatalf("events = %#v", got)
	}
	// The clean end is sticky: a caller that polls again keeps getting EOF
	// rather than an interruption.
	if _, err := stream.Next(); !errors.Is(err, io.EOF) {
		t.Fatalf("Next() after EOF error = %v, want io.EOF", err)
	}
}

func TestMessageStreamDoneInTrailingFrameStillEndsCleanly(t *testing.T) {
	t.Parallel()
	// No blank line after the last frame: the body ends mid-frame and the
	// done has to be projected from what was buffered.
	stream := testMessageStream(
		"event: chunk\ndata: {\"content\":\"answer\"}\n\n"+
			"event: done\ndata: {\"done\":true,\"usage\":{\"input_tokens\":1,\"output_tokens\":1}}",
		1024*1024,
		128*1024,
	)
	defer stream.Close()
	if event, err := stream.Next(); err != nil || event.Type != EventChunk {
		t.Fatalf("first Next() event=%#v err=%v", event, err)
	}
	if event, err := stream.Next(); err != nil || event.Type != EventDone {
		t.Fatalf("second Next() event=%#v err=%v", event, err)
	}
	if _, err := stream.Next(); !errors.Is(err, io.EOF) {
		t.Fatalf("third Next() error = %v, want io.EOF", err)
	}
}

// A connection dropped after a done but before the next one is
// indistinguishable from a clean end: the text that arrived is delivered and
// the body ends with io.EOF. Documented as the accepted trade-off in
// docs/api/openfang-gateway-consumption.md; pinned here so it cannot change
// silently.
func TestMessageStreamDeliversTextAfterTheLastDoneThenEOF(t *testing.T) {
	t.Parallel()
	stream := testMessageStream(
		"event: done\ndata: {\"done\":true,\"usage\":{\"input_tokens\":1,\"output_tokens\":1}}\n\n"+
			"event: chunk\ndata: {\"content\":\"late\"}\n\n",
		1024*1024,
		128*1024,
	)
	defer stream.Close()
	if event, err := stream.Next(); err != nil || event.Type != EventDone {
		t.Fatalf("first Next() event=%#v err=%v", event, err)
	}
	if event, err := stream.Next(); err != nil || event.Content != "late" {
		t.Fatalf("second Next() event=%#v err=%v", event, err)
	}
	if _, err := stream.Next(); !errors.Is(err, io.EOF) {
		t.Fatalf("third Next() error = %v, want io.EOF", err)
	}
}

func TestMessageStreamEOFAfterEventsButBeforeAnyDoneIsInterrupted(t *testing.T) {
	t.Parallel()
	stream := testMessageStream(
		"event: chunk\ndata: {\"content\":\"working\"}\n\n"+
			"event: tool_use\ndata: {\"tool\":\"shell\"}\n\n"+
			"event: phase\ndata: {\"phase\":\"tool_loop\",\"detail\":null}\n\n",
		1024*1024,
		128*1024,
	)
	defer stream.Close()
	for range 3 {
		if _, err := stream.Next(); err != nil {
			t.Fatalf("Next() error = %v", err)
		}
	}
	_, err := stream.Next()
	var streamErr *StreamError
	if !errors.As(err, &streamErr) || streamErr.Kind != StreamInterrupted {
		t.Fatalf("Next() error = %#v, want interrupted", err)
	}
}

func TestMessageStreamRejectsMalformedAndOversizedFrames(t *testing.T) {
	t.Parallel()
	for _, test := range []struct {
		name string
		body string
		max  int
		kind StreamErrorKind
	}{
		{
			name: "malformed json",
			body: "event: chunk\ndata: not-json\n\n",
			max:  1024,
			kind: StreamMalformed,
		},
		{
			name: "oversized event",
			body: "event: chunk\ndata: {\"content\":\"" + strings.Repeat("x", 1024) + "\"}\n\n",
			max:  128,
			kind: StreamLimit,
		},
	} {
		test := test
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()
			stream := testMessageStream(test.body, 4096, test.max)
			defer stream.Close()
			_, err := stream.Next()
			var streamErr *StreamError
			if !errors.As(err, &streamErr) || streamErr.Kind != test.kind {
				t.Fatalf("Next() error = %#v, want kind %s", err, test.kind)
			}
		})
	}
}

func testMessageStream(body string, maxTotal int64, maxEvent int) *MessageStream {
	ctx, cancel := context.WithCancel(context.Background())
	return newMessageStream(
		ctx,
		cancel,
		io.NopCloser(strings.NewReader(body)),
		"upstream-request",
		maxTotal,
		maxEvent,
	)
}
